Fix to `Pattern.or` and `Pattern.replicate` to properly do backtracking when given composite patterns

pchiusano opened the pull request:
Fixes #3530 thanks to @dolio for pairing on this!

Previously, patterns like:

```
Pattern.or (Pattern.join [p1, p2]) p3
```

Would not correctly backtrack the state (including both captures and the position) if `p1` succeeded. `Pattern.replicate` had a similar issue.

This PR fixes that and adds a regression (unit) test for the specific case reported by @ceedubs and a number of other related test cases.

I'm shocked this wasn't noticed before, but I'm guessing it's a combination of:

* Many "in the wild" patterns only need "1 token of lookahead" and are written as such. (Including examples in the docs and the prior unit tests 😳)
* API hasn't gotten a ton of usage yet

Note that if you do `capture (optional p)`, that will always capture at least the empty string (if `p` fails, then `optional` succeeds, consuming 0 characters, so the empty string consisting of 0 characters is captured), so the original bug report was slightly off in its expectations.

pchiusano committed:
unit test and wip

pchiusano committed:
fixup `Or` case

pchiusano committed:
tests passing, but needs some more tests

pchiusano committed:
more tests

pchiusano committed:
nested or test

pchiusano committed:
cleanup and add some comments

ceedubs reviewed (approved):
I won't claim to understand all of the changes or their implications, but the tests look good to me.

It would be nice if there were a way to test this closer to the Unison code. Is it possible to test `Pattern` via a transcript?

pchiusano committed:
added transcript regression test

pchiusano commented:
@ceedubs I went ahead and added a transcript regression test as well. :) We do also have transcript integration tests for the pattern stuff, so I added an example there:

https://github.com/unisonweb/unison/pull/3540/files#diff-9766e8bd8574eb3b1a896c857f18b737b01f11ab9c43082f3a28dcdd0e94ad00

ceedubs reviewed (approved):
Nice thanks @pchiusano!
